src/common/filemanager.cpp

```diff
@@ -0,0 +1,531 @@
+{
+	try {
+		path filepath = FileManager::userDataPath() / relative_path;
+		if (!FileManager::makeDirsForFile(filepath))
+			return false;
+
+		return impl::writeAbsPathFile(filepath, data, size);
+	}
+	catch (const std::bad_alloc &e) {
+		voxen::Log::warn("Out of memory: {}", e.what());
+		return false;
+	}
+}
+
+optional<string> FileManager::readUserTextFile(const path& relative_path) noexcept
+{
+	if (relative_path.empty())
+		return std::nullopt;
+
+	try {
+		impl::StringStorage storage;
+		bool success = impl::readAbsPath(FileManager::userDataPath() / relative_path, storage);
+		return success ? optional(storage.m_string) : std::nullopt;
+	}
+	catch (const std::bad_alloc &e) {
+		voxen::Log::warn("Out of memory: {}", e.what());
+		return std::nullopt;
+	}
+}
+
+bool FileManager::writeUserTextFile(const path& relative_path, const string& text) noexcept
+{
+	try {
+		path filepath = FileManager::userDataPath() / relative_path;
+		if (!FileManager::makeDirsForFile(filepath))
+			return false;
+
+		return impl::writeAbsPathFile(filepath, text.data(), text.size());
+	}
+	catch (const std::bad_alloc &e) {
+		voxen::Log::warn("Out of memory: {}", e.what());
+		return false;
+	}
+}
+
+optional<dyn_array<std::byte>> FileManager::readFile(const path& relative_path) noexcept
+{
+	if (relative_path.empty())
+		return std::nullopt;
+
+	try {
+		impl::RawBytesStorage storage;
+		bool success = impl::readAbsPath(FileManager::gameDataPath() / relative_path, storage);
+		return success ? optional(dyn_array<std::byte>(storage.m_data, storage.m_size)) : std::nullopt;
+	}
+	catch (const std::bad_alloc &e) {
+		voxen::Log::warn("Out of memory: {}", e.what());
+		return std::nullopt;
+	}
+}
+
+optional<string> FileManager::readTextFile(const path& relative_path) noexcept
+{
+	if (relative_path.empty())
+		return std::nullopt;
+
+	impl::StringStorage storage;
+	bool success = impl::readAbsPath(FileManager::gameDataPath() / relative_path, storage);
+	return success ? optional(storage.m_string) : std::nullopt;
+}
+
+bool FileManager::makeDirsForFile(const std::filesystem::path& relative_path) noexcept
+{
+	try {
+		std::filesystem::create_directories(relative_path.parent_path());
+	}
+	catch (const std::bad_alloc &e) {
+		voxen::Log::warn("Out of memory: {}", e.what());
+		return false;
+	}
+	catch (const std::filesystem::filesystem_error &e) {
+		voxen::Log::warn("Directory creation error: {}", e.what());
+		return false;
+	}
+
+	return true;
+}
+
+path FileManager::userDataPath() noexcept
+{
+	if constexpr (BuildConfig::kIsDeployBuild) {
+		return std::filesystem::path(sago::getDataHome()) / "voxen/";
+	} else {
+		return std::filesystem::current_path() / "data/user" / profile_name;
+	}
+}
+
+path FileManager::gameDataPath() noexcept
+{
+	if constexpr (BuildConfig::kIsDeployBuild) {
+		// TODO: implement me
+		assert(false);
+		return std::filesystem::path();
+	} else {
+		return std::filesystem::current_path() / "data/game";
+	}
+}
```

Comment:
These functions are technically not noexcept as `std::filesystem::current_path()` or `operator /` may throw.

Reply:
Done